File: algorithm/src/baekjoon/java/dynamic/BOJ11053.java
package baekjoon.java.dynamic;

import java.io.*;
import java.util.*;

public class BOJ11053 {
    private static int[] arr;
    private static int[] dp;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        arr = new int[n+1];
        dp = new int[n+1];
        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 1; i <= n; i++) arr[i] = Integer.parseInt(st.nextToken());

        int idx = 0;
        for (int i = 1; i <= n; i++) {
            int tmpIdx = idx;
            while (tmpIdx > 0 && arr[i] <= dp[tmpIdx]) tmpIdx--;
            if (dp[tmpIdx + 1] > arr[i]) dp[tmpIdx + 1] = arr[i];
            if (tmpIdx == idx) dp[++idx] = arr[i];
        }
        System.out.println(idx);
    }
}

File: algorithm/src/baekjoon/kotlin/boj/BOJ11053.kt
package baekjoon.kotlin.boj

import java.io.*
import java.util.StringTokenizer

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val n = readLine()!!.toInt()
    val st = StringTokenizer(readLine())
    val arr = IntArray(n) { st.nextToken().toInt() }
    val result = IntArray(n + 1)
    var max = 1
    result[max] = arr[0]
    for (i in 1 until n) {
        var k = max
        while (arr[i] <= result[k]) k--
        if (k <= 0) result[1] = arr[i]
        else {
            result[k + 1] = arr[i]
            if (k == max) max++
        }
    }
    println(max)
}
